### DISCUSSION TITLE
Some JSON column data gets saved as a string

### DISCUSSION THREAD

**Discussion by MrBrax** [prisma/prisma#20903] [Q&A] [answered]
### Question

I have this data that is sent to a create/upsert operation, part of a larger root JSON object:
```
"inventory": {
        "i": "a0585dde-5e80-4b86-80b4-354c71c83c33",
        "s": [
            {
                "iid": "a0585dde-5e80-4b86-80b4-354c71c83c33",
                "ix": 0,
                "t": "Furniture",
                "i": {
                    "$type": "furniture",
                    "furnitureId": "marbodal.drawers",
                    "lightGroup": 1,
                    "lightEnabled": false,
                    "entityType": "Furniture"
                },
                "a": 1
            }
        ],
        "w": 8,
        "h": 2
    }
```

However, when it is inserted into the database, it looks like this:
```
{
  h: 2,
  i: 'a0585dde-5e80-4b86-80b4-354c71c83c33',
  s: [
    {
      a: 1,
      t: 'Furniture',
      ix: 0,
      iid: 'a0585dde-5e80-4b86-80b4-354c71c83c33',
      i: '{"$type":"furniture","furnitureId":"marbodal.drawers","lightGroup":1,"lightEnabled":false,"entityType":"Furniture"}'
    }
  ],
  w: 8
}
```

Why is this happening? The correct structure is kept until it is inserted into the database, then `item` is just serialized again inside the JSON object. Is this a Postgres or Prisma issue?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model Player {
inventory Json @default("{}")
}
```

```ts
player = await getPrisma().player.upsert({
            where: {
                steamid64: req.params.id,
            },
            update: data,
            create: data,
        });
```

- OS: Windows
- Database: PostgreSQL
- Node.js version: v20.5.0

- Run `prisma -v` to see your Prisma version and paste it
prisma                  : 5.1.1
@prisma/client          : 5.2.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\query_engine-windows.dll.node)Schema Engine           : schema-engine-cli 6a3747c37ff169c90047725a05a6ef02e32ac97e (at node_modules\@prisma\engines\schema-engine-windows.exe)  
Schema Wasm             : @prisma/prisma-schema-wasm 5.1.1-1.6a3747c37ff169c90047725a05a6ef02e32ac97e
Default Engines Hash    : 6a3747c37ff169c90047725a05a6ef02e32ac97e
Studio                  : 0.492.0

**Comment by nurul3101** [comment]
Hello @MrBrax 👋 

If I understand correctly, I can see that the order of keys in the JSON object were re-arranged after insertion and you would have expected the keys to be in the same order as you passed while insertion, right?

Could you check if you get the same behaviour while inserting a record through [Raw Prisma query](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access) as well?

**Reply by MrBrax** [reply]
No, the problem is that the "i" value inside the "s" array is turned into a string when put into the database, it is no longer a JSON object.

I could try a raw query.

**Reply by MrBrax** [reply]
@nurul3101 Indeed, when using a raw prisma query the entire JSON object is untouched.

**Comment by MrBrax** [comment]
After some more logging i found something out:

inventory before prisma with console.log:
```
{
  i: '2eb5473d-0e25-42e1-a19c-262abb428bf6',
  s: [
    {
      iid: '2eb5473d-0e25-42e1-a19c-262abb428bf6',
      ix: 0,
      t: 'Furniture',
      item: [Object],
      a: 1
    }
  ],
  w: 8,
  h: 2
}
```

inventory before prisma with console.log + json.stringify:
```
{
  "i": "2eb5473d-0e25-42e1-a19c-262abb428bf6",
  "s": [
    {
      "iid": "2eb5473d-0e25-42e1-a19c-262abb428bf6",
      "ix": 0,
      "t": "Furniture",
      "item": {
        "$type": "furniture",
        "furnitureId": "table",
        "lightGroup": 0,
        "lightEnabled": false,
        "entityType": "Furniture"
      },
      "a": 1
    }
  ],
  "w": 8,
  "h": 2
}
```

inventory after prisma insert, console.log the return object from upsert:
```
{
  h: 2,
  i: '2eb5473d-0e25-42e1-a19c-262abb428bf6',
  s: [
    {
      a: 1,
      t: 'Furniture',
      ix: 0,
      iid: '2eb5473d-0e25-42e1-a19c-262abb428bf6',
      item: '{"$type":"furniture","furnitureId":"table","lightGroup":0,"lightEnabled":false,"entityType":"Furniture"}'
    }
  ],
  w: 8
}
```

Notice how `item` is logged as `[Object]` when not using json.stringify? Is prisma using some other internal method when inserting?

**Comment by MrBrax** [comment, marked as answer]
Found the issue. `$type` is used internally in Prisma, which i haven't encountered in the docs. I changed the discriminator in C# and it now works fine.

https://github.com/prisma/prisma/blob/62b4e564b436e2e42853ce42cddfbd3dfdb7d891/packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts#L254
